Group prometheus metrics

henriquegemignani opened the pull request:
These values are provided via labels and it's much easier when handling Prometheus data to filter via a label than to merge data from two different metrics.
This allows a dashboard to be generic enough to work for different machines.

Labels with data about which machine it is, and which job provided the data are added by Prometheus itself.

nickbabcock commented:
Thank you for the pull request, looking good!

I'm not a daily prometheus user (as is probably evident), and this pull request got me thinking that the metrics we expose need a bit of love in order to follow [Prometheus's best practices](https://prometheus.io/docs/practices/naming/):

> [A metric name] should have a (single-word) application prefix relevant to the domain the metric belongs to. For metrics specific to an application, the prefix is usually the application name itself.

**Before PR**: No application namespace

**After PR**: Uses the `ohm` namespace for all metrics produced by OhmGraphite

✔️ good change, brings us in line with the recommendation of using a namespace for applications. Let's keep this.

> [A metric name] should use base units (e.g. seconds, bytes, meters - not milliseconds, megabytes, kilometers).

I believe we can adhere to this better:

Todo: depending on `SensorType`, data will need to be transformed:

- `SensorType.Clock`: MHz => Hz
- `SensorType.Data`: GB (2^30) => Bytes
- `SensorType.SmallData`: MB (2^20) => Bytes
- etc

Todo: Depending on the sensor type, it should be translated to use base units when creating the metric name:

- `SensorType.Temperature`: celsius? Though simply "temperature" doesn't seem too unreasonable
- `SensorType.Voltage`: volts
- `SensorType.Level / Load / Control`: percent converted into ratio?
- etc

> [A metric name] must comply with the [data model](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels) for valid characters.

We should use the official regex implementation to remove any non-conforming characters.

---

Seeing the prometheus docs show metric names like "node_memory_usage_bytes" and "process_cpu_seconds_total" do you think that something like "ohm_temperature" is too general? Should it include the component's name (like "ohm_cpu_temperature"?). To lend more support to this rationale, the [official collectd plugin](https://github.com/prometheus/collectd_exporter/blob/6bab6e910e0012f0909f07a0c611fc43b57af59c/main.go#L61-L77) incorporates the plugin name (eg: cpu, disk, impi, memory) when forming metric names.

---

What we're proposing is a pretty major breaking change for any prometheus user exploring OhmGraphite data, which is ok as breaking changes happen -- I'll just need to explain it in the future release notes why a breaking change was the right thing for us to do.

---

Sorry this has turned into a bit of a ramble, but how do you feel about taking this PR one step farther:

- Incorporate HardwareType into the metric name (and remove it as a label)
- Thus the metric name would be `ohm_<HardwareType>_<SensorType>` (eg: "ohm_cpu_hz")
- Remove sensor index as a label as keeping `Sensor` as label should already distinguish between multiple sensors on the same hardware (eg: a sensor name of `Voltage #12` makes storing sensor_index=11 seem wasteful) 
- Thus the only labels are: `Hardware` and `Sensor`

henriquegemignani commented:
Heya, thanks for the feedback! :)

I've just started using Prometheus myself and the way the metrics were being exported were preventing me from tracking multiple machines at the same time.

I added the hardware type to the metric name and added the sensor type as well. For some of the sensor types I've added the unit, but some I don't really know the proper one.

nickbabcock commented:
Are you comfortable with me merging this and I can continue working on improving prometheus support? Thank you so much for pitching in 😄

henriquegemignani commented:
Sure! If you need any assistance, feel free to ask.
